Make sure 'final' decls never get marked as 'open'

jrose-apple opened the pull request:
Follow-up to #15996:

- Mark `willSet`/`didSet` accessors as `private`, always. These functions are always immediately inlined into a synthesized setter, and cannot be used with computed properties, so there's no benefit (and some harm) in having their access follow the property's.

- Disallow `open let`, and add a verifier check to make sure we don't mix `open` and `final` in the future. You could *write* `open let` in Swift 4.1, but you couldn't actually override such a property (within the module or outside). To maintain source compatibility, this is just going to emit a warning before Swift 5 (instead of an error), since it wasn't harming anything in practice.

I wouldn't have bothered with the first part right now except that it was the only other place where we ended up with declarations that were both `final` and `open` under the old code.

jrose-apple committed:
Mark willSet/didSet as 'private', always

These functions are always immediately inlined into a synthesized
setter, and cannot be used with computed properties, so there's no
benefit (and some harm) in having their access follow the property's.

No intended user-visible semantics change.

jrose-apple committed:
Disallow 'open let'

...and add a verifier check to make sure we don't mix 'open' and
'final' in the future.

You could /write/ 'open let' in Swift 4.1, but you couldn't actually
override such a property (within the module or outside). To maintain
source compatibility, this is just going to emit a warning before
Swift 5 (instead of an error), since it wasn't harming anything in
practice.

jrose-apple commented:
@swift-ci Please test

jrose-apple commented:
@swift-ci Please test source compatibility

jrose-apple commented:
@swift-ci Please test Linux

slavapestov reviewed (approved):
Thanks for tightening up invariants!

jrose-apple commented:
Hm, looks like I have more work to do!

jrose-apple committed:
Also check for `open static` and 'open'-in-'final'-class

Like the change for `open let`, this is only a warning in Swift 4
mode, because it didn't actually cause any harm in practice. Neither a
Swift developer nor a compiler developer wants to rely on that going
forward, though!

jrose-apple committed:
Simplify inferDynamic by assuming 'final' has already been inferred

...and then reorder calls so that that is always the case.

No intended functionality change.

jrose-apple commented:
Two additional commits. @slavapestov, mind glancing it over once more?

@swift-ci Please test

jrose-apple commented:
@swift-ci Please test source compatibility

slavapestov commented on lib/Sema/TypeCheckDecl.cpp:
```diff
@@ -1223,6 +1224,57 @@ void swift::makeDynamic(ASTContext &ctx, ValueDecl *D) {
+    if (auto finalAttr = D->getAttrs().getAttribute<FinalAttr>()) {
+      auto finalRange = finalAttr->getRange();
+      if (finalRange.isValid()) {
+        TC.diagnose(finalRange.Start, diag::static_decl_already_final)
```
I think we have a diagnoseAndRemoveAttr() utility somewhere

jrose-apple replied:
So we do…but currently only in TypeCheckAttr.cpp. I think I've done enough moving little things around for the moment.

slavapestov replied:
No worries. There are many yaks to shave here and the PR is already a solid improvement.
